include/ofi_mem.h

```diff
@@ -289,6 +289,7 @@ enum {
 	OFI_BUFPOOL_INDEXED		= 1 << 1,
 	OFI_BUFPOOL_NO_TRACK		= 1 << 2,
 	OFI_BUFPOOL_HUGEPAGES		= 1 << 3,
+	OFI_BUFPOOL_MMAP		= 1 << 4,
```

Comment:
I'm saying that HUGEPAGES implies MMAP.  That's how huge pages are allocated.  So, these aren't separate.  It's the reverse that's not true (MMAP -/-> HUGEPAGES).  That's why I would define HUGEPAGES = MMAP | BIT(x).

Then the code path below would automatically fall back from huge pages to mmap, which guarantees page aligned allocations, which is closer to the behavior the caller would have gotten specifying HUGEPAGES.  A user can still set MMAP without HUGEPAGES.

Reply:
I understand the argument, but I think this will lead to brittle code, and I am not sure that the benefits outweigh this cost.  Testing for HUGEPAGES would now need to also be aware of the special case.  Instead of 
```
if (pool->attr.flags & OFI_BUFPOOL_HUGEPAGES) {
}
if (pool->attr.flags & OFI_BUFPOOL_MMAP) {
}
```
one would now have to write:
```
if (pool->attr.flags & (OFI_BUFPOOL_HUGEPAGES & ~OFI_BUFPOOL_MMAP)) {
}
if (pool->attr.flags & OFI_BUFPOOL_MMAP) {
}
```
otherwise `flags == OFI_BUFPOOL_MMAP` would match the first `if` statement (that expression is non-zero).

While I can do this for this review, I fear that this is just creating a footgun for some future maintainer not aware of this conversation.

That said, I am open to making the change.  If the decision is to combine the bits it would probably be cleaner to provide some predicate functions like `bufpool_uses_hugepages` and `bufpool_uses_mmap` to encapsulate the ugliness above.

Reply:
Hmm... that's a good point.  The real objective from the viewpoint of the user is to obtain page aligned allocations.  Maybe the implementation stays as proposed, and we really just need a better name for the flag.  It really doesn't matter if we use mmap (which is the easy implementation) or a more complex algorithm with malloc.  The page alignment must hold true, and it can't safely fallback to a standard malloc without silently causing issues (memory corruption in this case).  The hugepage flag is basically defined as a preference (and maybe should have been called HUGEPAGES_PREFERRED). That makes it clear that non-huge pages are acceptable.  But this flag is really forcing a requirement that must be maintained (PAGE_ALIGNED?).  Keeping the flag still keeps the patch simple without going to the extreme of trying to support a generic alloc_alignment.

Reply:
Thinking about this more, the requirement is actually `mmap` rather than page alignment.  Imagine that we use `posix_memalign` to obtain a page-aligned buffer and called `ibv_reg_mr` on it, setting `MADV_DONTFORK` as a side effect.  Then we fork, and in the child these pages are not mapped.  Even if the child can successfully call `free()` on this chunk, it'll just explode the next time malloc tries to reuse this memory with a hole in it.

In summary: it is never safe to use malloc for these allocations and they must come from a separate mapping, so I think we should call this flag MMAP.

Reply:
@nzmsv I think what you said make sense. The main concern I have is the fallback behavior for huge page.

Because huge page implies using mmap, when huge page allocation failed, it should fall back to  mmap.

I suggest to add the  OFI_BUFPOOL_MMAP flag when huge page allocation fail.

Reply:
@wzamazon I am adding the flag here: https://github.com/nzmsv/libfabric/pull/1

Not adding the flag by default gives more flexibility to the caller (they can choose to fall back to mmap or small pages) but I suppose it's another chance to make a mistake.  I'm OK with either solution.

Reply:
I added the flag to the fallback from hugepages to mmap.  Currently we'll only fall back from mmap to malloc if the error returned is `FI_ENOSYS` (so we are on an OS that does not support mmap).